app/controllers/bookmarks_controller.rb

```diff
@@ -0,0 +1,63 @@
+class BookmarksController < ApplicationController
+  def show
+  end
+
+  def new
+    @user = current_user
+    @topic = @user.topics.find(params[:topic_id])
+    @bookmark = @topic.bookmarks.new
+  end
+  
+  def create
+    @user = current_user
+    @topic = @user.topics.find(params[:topic_id])
+    @bookmark = @topic.bookmarks.build(bookmark_params)
+    
+    if @bookmark.save
+      flash[:notice] = "Bookmark created successfully"
+      redirect_to(@topic)
+    else
+      flash[:alert] = "Issue creating bookmark"
+      render :new
+    end
+  end
+
+  def edit
+    @user = current_user
+    @topic = @user.topics.find(params[:topic_id])
+    @bookmark = Bookmark.find(params[:id])
+  end
+  
+  def update
```

Comment:
@ianroberts131 Currently a logged-in attacker could edit & update the bookmarks belonging to a victim user if they are able to guess a bookmark id and topic id that belongs to the victim.

*Before* fixing the vulnerability, try to write a *controller* spec that tests an attacker cannot update a bookmark belonging to a victim user. See the test fail in an expected way to prove the vulnerability exists.

*After* writing the spec and seeing it fail expectedly, then try to fix the vulnerability. You'll know the vulnerability is fixed once the test passes.

Finally commit the spec and fix to your repo. This spec now exists as a future guard against the vulnerability accidentally reappearing in the future, say when we make further changes to this controller.

Try to cultivate the habit of running all of your specs before merging branches into master. All an app's specs should pass before merging into master. This helps maintain the quality of the master branch.